src/rsa.rs

```diff
@@ -163,8 +165,107 @@ fn parse_public_key<'a>(input: untrusted::Input<'a>) ->
     })
 }
 
+#[allow(non_camel_case_types)]
+enum BIGNUM {}
+#[allow(non_camel_case_types)]
+enum BN_MONT_CTX {}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct RSA {
+    n: *mut BIGNUM,
+    e: *mut BIGNUM,
+    d: *mut BIGNUM,
+    p: *mut BIGNUM,
+    q: *mut BIGNUM,
+    dmp1: *mut BIGNUM,
+    dmq1: *mut BIGNUM,
+    iqmp: *mut BIGNUM,
+    mont_n: *mut BN_MONT_CTX,
+    mont_p: *mut BN_MONT_CTX,
+    mont_q: *mut BN_MONT_CTX,
+    mont_qq: *mut BN_MONT_CTX,
+    qmn_mont: *mut BIGNUM,
+    iqmp_mont: *mut BIGNUM,
+}
+
+/// An RSA key pair.
+pub struct RSAKeyPair {
+    rsa: Box<RSA>,
+}
+
+const BN_FLG_CONSTTIME: c::int = 4;
+
+fn build_bn(input: untrusted::Input) -> *mut BIGNUM {
+    unsafe {
+        let bytes = input.as_slice_less_safe();
+        let res = BN_bin2bn(bytes.as_ptr(), bytes.len(), ptr::null_mut());
+        BN_set_flags(res, BN_FLG_CONSTTIME);
+        res
+    }
+}
+
+impl RSAKeyPair {
+    /// Reads key data from the given DER-encoded input, which must contain
+    /// the full private key data (n, e, d, p, q, dmp1, dmq1, iqmp).
+    ///
+    /// This is intended for use by code that deserializes key pairs.
+    pub fn from_der(input: untrusted::Input) -> Result<RSAKeyPair, ()> {
+        input.read_all((), |input| {
+            der::nested(input, der::Tag::Sequence, (), |input| {
+                let _ = try!(der::small_nonnegative_integer(input)); // version
```

Comment:
You need to check that the version is zero and fail if it isn't zero. Please add four new test vectors for this: One where the version field is missing, one where version is 1, and one where the version is -1 (encoded as value 0xff), and one where the version is 2. (Please create these test vectors by taking an existing good test vector and modifying it.)